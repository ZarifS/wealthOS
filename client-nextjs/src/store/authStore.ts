import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import { getLocalStorageWithExpiry } from 'utils';
import AuthService, { RegisterPayload, LoginPayload, TOKEN_KEY } from '../services/authService';

let token;

if (typeof window !== 'undefined') {
  token = getLocalStorageWithExpiry(TOKEN_KEY);
}

interface AuthState {
  isLoggedIn: boolean;
  loading: boolean;
  token: string | null;
  message: string | null;
}

const initialState: AuthState = token
  ? { isLoggedIn: true, token, loading: false, message: null }
  : { isLoggedIn: false, token: null, loading: false, message: null };

export const register = createAsyncThunk(
  'auth/register',
  async (
    { firstName, lastName, email, password, confirmedPassword }: RegisterPayload,
    thunkAPI
  ) => {
    try {
      const { token } = await AuthService.register({
        firstName,
        lastName,
        email,
        password,
        confirmedPassword,
      });
      return { token };
    } catch (error: any) {
      const message =
        (error.response && error.response.data && error.response.data.message) ||
        error.message ||
        error.toString();
      return thunkAPI.rejectWithValue(message);
    }
  }
);

export const login = createAsyncThunk(
  'auth/login',
  async ({ email, password }: LoginPayload, thunkAPI) => {
    try {
      const { token } = await AuthService.login({ email, password });
      return { token };
    } catch (error: any) {
      const message =
        (error.response && error.response.data && error.response.data.message) ||
        error.message ||
        error.toString();
      return thunkAPI.rejectWithValue(message);
    }
  }
);


const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    logout: (state) => {
      AuthService.logout();
      state.isLoggedIn = false;
      state.token = null;
      state.loading = false;
      state.message = null;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(register.pending, (state: AuthState) => {
      state.loading = true;
      state.message = null;
      state.isLoggedIn = false;
      state.token = null;
    });
    builder.addCase(register.fulfilled, (state: AuthState, action: PayloadAction<any>) => {
      state.isLoggedIn = true;
      state.token = action.payload.token;
      state.loading = false;
    });
    builder.addCase(register.rejected, (state: AuthState, action: PayloadAction<any>) => {
      state.loading = false;
      state.message = action.payload;
    });
    builder.addCase(login.pending, (state: AuthState) => {
      state.loading = true;
      state.message = null;
      state.token = null;
      state.isLoggedIn = false;
    });
    builder.addCase(login.fulfilled, (state: AuthState, action: PayloadAction<any>) => {
      state.isLoggedIn = true;
      state.token = action.payload.token;
      state.loading = false;
    });
    builder.addCase(login.rejected, (state: AuthState, action: PayloadAction<any>) => {
      state.loading = false;
      state.message = action.payload;
    });
  },
});

const { reducer } = authSlice;
export default reducer;
