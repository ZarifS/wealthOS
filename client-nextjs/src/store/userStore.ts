// userStore.ts
import { createSlice, createAsyncThunk, PayloadAction } from '@reduxjs/toolkit';
import {getUserData} from '../services/userService';

export interface User {
    uuid: string;
    firstName: string;
    lastName: string;
    email: string;
    spaces: [string];
}

interface UserState {
    user: User | null;
    loading: boolean;
    error: {status?: number, message: string} | null;
  }

// Define the initial state
const initialState: UserState = {
  user: null,
  loading: false,
  error: null,
};

// Create async thunk for fetching user data
export const fetchUserData = createAsyncThunk(
  'user/fetchUserData',
  async (token: string, thunkApi) => {
    try {
      const response = await getUserData(token);
      return response.data;
    } catch (error: any) {
      const message =
      (error.response && error.response.data && error.response.data.message) ||
        error.message || error.toString();
      console.log('Error trying to fetch user data:', message);
      return thunkApi.rejectWithValue(message);
    }
  }
);

// Create the slice
const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchUserData.pending, (state: UserState) => {
        state.loading = true;
        state.error = null;
      })
      .addCase(fetchUserData.fulfilled, (state: UserState, action: PayloadAction<any>) => {
        state.loading = false;
        state.user = action.payload;
      })
      .addCase(fetchUserData.rejected, (state: UserState, action: PayloadAction<any>) => {
        state.loading = false;
        state.error = {message: action.payload};
      });
  },
});

export default userSlice.reducer;
