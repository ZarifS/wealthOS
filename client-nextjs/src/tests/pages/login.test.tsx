import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../store';
import Login from '../../pages/login';

describe('Login page', () => {
    const setup = () => {
        return render(
            <Provider store={store}>
                <Login />
            </Provider>
        );
    };

    it('renders correctly', () => {
        const { asFragment } = setup();
        expect(asFragment()).toMatchSnapshot();
    });

    it('submits the form', async () => {
        setup();

        fireEvent.input(screen.getByLabelText('Email'), {
            target: { value: 'john@example.com' },
        });
        fireEvent.input(screen.getByLabelText('Password'), {
            target: { value: 'password123' },
        });

        fireEvent.click(screen.getByRole('button', { name: 'Sign In' }));

        await waitFor(() =>
            expect(screen.queryByText('Loading...')).toBeInTheDocument()
        );
    });
});
