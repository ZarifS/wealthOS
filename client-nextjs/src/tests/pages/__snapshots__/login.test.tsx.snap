// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page renders correctly 1`] = `
<DocumentFragment>
  <div
    class="main"
  >
    <h1
      class="header"
    >
      Sign In
    </h1>
    <form
      class="form"
    >
      <div
        class="formInputs"
      >
        <div
          class="main"
        >
          <label
            class="label"
            for="email"
          >
            Email
          </label>
          <input
            class="input"
            id="email"
            name="password"
            placeholder="john.doe@gmail.com"
            type="email"
            value=""
          />
        </div>
        <div
          class="main"
        >
          <label
            class="label"
            for="password"
          >
            Password
          </label>
          <input
            class="input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
      </div>
      <div
        class="message"
      />
      <button
        class="customButton "
        type="submit"
      >
        Sign In
      </button>
    </form>
    <div
      class="switchAuth"
    >
      <a
        href="/register"
      >
        <p>
          Don't have an account? Register now!
        </p>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
